### PR TITLE
Updates references to v1alpha4 types

### PR DESCRIPTION
**Pull request by srm09**
**What this PR does / why we need it**:
The CAPI module in use now points to the latest on the `master` branch of CAPI. This patch updates the references from `v1alpha3` to `v1alpha4` types.
This patch has the following changes:
- Use the latest `master` from the CAPI repo.
- Updates controller-runtime to version `v0.8.2`
- Updates golang version to `v1.15.8`
- Updates the CAPV codebase to reference v1alpha4 types.
- Uses the CAPI nightly builds for E2E tests with v1alpha4 types

**Which issue(s) this PR fixes**:
n/a

**Special notes for your reviewer**:
n/a

**Release note**:
```release-note
Introduces support for v1alpha4 types
```

### CODE REVIEW

**Comment by srm09** [comment]
@yastij Does it make sense to break this into two separate PRs
- One with the reference updates for `v1alpha4`
- And the other one with just the test setup and the single test

**Comment by srm09** [comment]
Re-arranged the commits to support breaking the PRs, if necessary

**Comment by srm09** [comment]
PR to update golang version in the prow jobs: https://github.com/kubernetes/test-infra/pull/21294

File: pkg/manager/manager.go [hunk 1]
```diff
@@ -73,7 +70,6 @@ func New(opts Options) (Manager, error) {
 		SyncPeriod:              &opts.SyncPeriod,
 		Namespace:               opts.WatchNamespace,
 		NewCache:                opts.NewCache,
-		NewClient:               newClientFunc,
```
Comment (srm09) [inline pkg/manager/manager.go:76]: This is not needed anymore, it is a default option with the newer version of `controller-runtime`

File: pkg/manager/manager.go [hunk 2]
```diff
@@ -118,20 +114,3 @@ type manager struct {
...
-	if err != nil {
-		return nil, err
-	}
-
-	return &client.DelegatingClient{
-		Reader:       cache,
-		Writer:       c,
-		StatusClient: c,
-	}, nil
-}
```
Comment (srm09) [inline pkg/manager/manager.go:137]: same as above

File: test/e2e/data/infrastructure-vsphere/kustomization/cluster-resource-set.yaml [hunk 3]
```diff
@@ -4,7 +4,7 @@ metadata:
   name: "cni-${CLUSTER_NAME}-crs-0"
 data: ${CNI_RESOURCES}
 ---
-apiVersion: addons.cluster.x-k8s.io/v1alpha3
```
Comment (srm09) [inline test/e2e/data/infrastructure-vsphere/kustomization/cluster-resource-set.yaml:7]: Am unsure about this one, is this necessary.

**Review by srm09 — COMMENTED** [review]
Added some comments about a couple of changes

**Comment by srm09** [comment]
@yastij This is ready for review now.

File: hack/tools/go.mod [hunk 4]
```diff
@@ -1,14 +1,14 @@
...
+go 1.15
 
 require (
 	github.com/golangci/golangci-lint v1.23.8
 	github.com/onsi/ginkgo v1.12.0
 	github.com/vmware/govmomi v0.23.1
-	k8s.io/code-generator v0.18.0
-	sigs.k8s.io/controller-tools v0.2.9
+	k8s.io/code-generator v0.20.2
+	sigs.k8s.io/controller-tools v0.4.1-0.20201002000720-57250aac17f6
```
Comment (gab-satchi) [inline hack/tools/go.mod:10]: does the controller-tools need to be the nightly build version as well? I see capi is using v0.5.0

File: pkg/context/fake/fake_cluster_context.go [hunk 5]
```diff
@@ -49,27 +49,27 @@ func NewClusterContext(ctx *context.ControllerContext) *context.ClusterContext {
 	}
 }
 
-func newClusterV1a2() clusterv1a2.Cluster {
-	return clusterv1a2.Cluster{
+func newClusterV1a2() clusterv1.Cluster {
```
Comment (gab-satchi) [inline pkg/context/fake/fake_cluster_context.go:52]: maybe drop the `a2` from this function now that it's on v1a4

File: pkg/context/fake/fake_controller_manager_context.go [hunk 6]
```diff
@@ -22,11 +22,11 @@ import (
 	"k8s.io/apimachinery/pkg/runtime"
 	clientgoscheme "k8s.io/client-go/kubernetes/scheme"
 	clientrecord "k8s.io/client-go/tools/record"
-	clusterv1a2 "sigs.k8s.io/cluster-api/api/v1alpha3"
+	clusterv1a2 "sigs.k8s.io/cluster-api/api/v1alpha4"
```
Comment (gab-satchi) [inline pkg/context/fake/fake_controller_manager_context.go:25]: can rename to `clusterv1`?

**Review by gab-satchi — COMMENTED** [review]
broadly lgtm. just minor nits

File: hack/tools/go.mod (hunk 4 above)
Comment (srm09) [reply hack/tools/go.mod:10]: Currently, we are pulling in the CAPI sha which was before CAPI moved to go1.16 and k8s1.21. Hence `controller-tools` points to a nightly build. We might have another change to move forward to the latest CAPI version once an early build for version 0.4 gets released.

File: main.go [hunk 7]
```diff
@@ -136,78 +135,46 @@ func main() {
...
-				return err
-			}
-
-			if err := (&v1alpha3.HAProxyLoadBalancer{}).SetupWebhookWithManager(mgr); err != nil {
-				return err
-			}
-			if err := (&v1alpha3.HAProxyLoadBalancerList{}).SetupWebhookWithManager(mgr); err != nil {
-				return err
-			}
-		} else {
```
Comment (yastij) [inline main.go:198]: I guess this is part of moving to the multi-tenancy model right ?

File: test/e2e/config/vsphere-ci.yaml [hunk 8]
```diff
@@ -28,10 +28,12 @@ providers:
   - name: cluster-api
     type: CoreProvider
     versions:
-      - name: v0.3.11
+      - name: v0.4.0
         # Use manifest from source files
-        value: "https://github.com/kubernetes-sigs/cluster-api/releases/download/v0.3.11/core-components.yaml"
+        value: "https://storage.googleapis.com/artifacts.k8s-staging-cluster-api.appspot.com/components/nightly_master_20210310/core-components.yaml"
```
Comment (yastij) [inline test/e2e/config/vsphere-ci.yaml:33]: this means we're moving our e2e to test against CAPI main-nightly, let's keep an eye on e2e if something breaks

File: test/e2e/data/shared/metadata.yaml [hunk 9]
```diff
@@ -0,0 +1,13 @@
+---
+apiVersion: clusterctl.cluster.x-k8s.io/v1alpha3
+kind: Metadata
```
Comment (yastij) [inline test/e2e/data/shared/metadata.yaml:3]: why do we need this ?

File: main.go (hunk 7 above)
Comment (srm09) [reply main.go:198]: Yes.

File: test/e2e/config/vsphere-ci.yaml (hunk 8 above)
Comment (srm09) [reply test/e2e/config/vsphere-ci.yaml:33]: Technically we are not moving to the latest nightly-build, and it is fixed to a particular SHA. I ran the E2E tests with this image, and they work. We might have to keep moving this regularly to the latest available nightly build.

File: test/e2e/data/shared/metadata.yaml (hunk 9 above)
Comment (srm09) [reply test/e2e/data/shared/metadata.yaml:3]: To make it aware of the versioning scheme in the latest CAPI nightly builds that we are pulling in.

**Comment by srm09** [comment]
@yastij Squashed the commits
/unhold

File: hack/tools/go.mod (hunk 4 above)
Comment (srm09) [reply hack/tools/go.mod:10]: Opening an issue to track this one

**Pull request merged**